lib/web_ui/lib/src/engine/canvaskit/text.dart

```diff
@@ -902,176 +905,21 @@ class CkParagraphBuilder implements ui.ParagraphBuilder {
-    final FontFallbackData fallbackData = FontFallbackData.instance;
-    codeUnits.removeWhere((int codeUnit) =>
-        fallbackData.knownCoveredCodeUnits.contains(codeUnit));
-    if (codeUnits.isEmpty) {
-      return true;
-    }
-
-    // We don't know if the remaining code units are covered by our fallback
-    // fonts. Check them and update the cache.
-    List<bool> codeUnitsSupported = List<bool>.filled(codeUnits.length, false);
-    String testString = String.fromCharCodes(codeUnits);
-
-    for (String font in fallbackData.globalFontFallbacks) {
-      List<SkTypeface>? typefacesForFamily =
-          skiaFontCollection.familyToTypefaceMap[font];
-      if (typefacesForFamily == null) {
-        printWarning('A fallback font was registered but we '
-            'cannot retrieve the typeface for it.');
-        continue;
-      }
-      for (SkTypeface typeface in typefacesForFamily) {
-        SkFont font = SkFont(typeface);
-        Uint8List glyphs = font.getGlyphIDs(testString);
-        assert(glyphs.length == codeUnitsSupported.length);
-        for (int i = 0; i < glyphs.length; i++) {
-          bool codeUnitSupported = glyphs[i] != 0;
-          if (codeUnitSupported) {
-            fallbackData.knownCoveredCodeUnits.add(codeUnits[i]);
-          }
-          codeUnitsSupported[i] |=
-              codeUnitSupported || _isControlCode(codeUnits[i]);
-        }
-      }
-
-      // Once we've checked every typeface for this family, check to see if
-      // every code unit has been covered in order to avoid unnecessary checks.
-      bool keepGoing = false;
-      for (bool supported in codeUnitsSupported) {
-        if (!supported) {
-          keepGoing = true;
-          break;
-        }
-      }
-
-      if (!keepGoing) {
-        // Every code unit is supported, clear [codeUnits] and return `true`.
-        codeUnits.clear();
-        return true;
-      }
-    }
-
-    // If we reached here, then there are some code units which aren't covered
-    // by the global fallback fonts. Remove the ones which were covered and
-    // return false.
-    for (int i = codeUnits.length - 1; i >= 0; i--) {
-      if (codeUnitsSupported[i]) {
-        codeUnits.removeAt(i);
-      }
-    }
-    return false;
-  }
-
-  /// Returns `true` if every code unit in [codeUnits] is known to not have any
-  /// fallback font which can cover it.
-  ///
-  /// This method has a side effect of removing every code unit from [codeUnits]
-  /// which is known not to have a fallback font which covers it.
-  bool _checkIfNoFallbackFontSupports(List<int> codeUnits) {
-    final FontFallbackData fallbackData = FontFallbackData.instance;
-    codeUnits.removeWhere((int codeUnit) =>
-        fallbackData.codeUnitsWithNoKnownFont.contains(codeUnit));
-    return codeUnits.isEmpty;
-  }
-
-  @override
-  void addText(String text) {
-    _ensureFontsSupportText(text);
+    FontFallbackData.instance.ensureFontsSupportText(text, fontFamilies);
```

Comment:
It's great that all the font-related stuff is moving out of the paragraph code. Separation of concerns FTW!